### PR TITLE
Mz/encapsulate degree

### PR DESCRIPTION
**Pull request by mayeul-zama**

### CODE REVIEW

**Review by IceTDrinker — COMMENTED** [review]
Have looked at the high level design for now to avoid spreading comments around

File: tfhe/src/shortint/ciphertext/mod.rs [hunk 1]
```diff
@@ -102,6 +98,56 @@ impl std::ops::Mul<usize> for NoiseLevel {
...
+    }
+
+    pub fn from_msg_carry_modulus(
+        msg_modulus: MessageModulus,
+        carry_modulus: CarryModulus,
+    ) -> Self {
+        Self(carry_modulus.0 * msg_modulus.0 - 1)
+    }
+
+    pub fn valid(&self, degree: Degree) -> Result<(), CheckError> {
```
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/shortint/ciphertext/mod.rs:121]: or maybe follow the noise level design for the naming maybe ? Unsure

[the remark this answers is not in the document]
Comment (tmontaigu) [reply tfhe/src/shortint/ciphertext/mod.rs:121]: I agree that seeing max_degree.valid(some_degree) seems weird, but its consistent with the way noise level is done, so if we change one, we should change both

[the remark this answers is not in the document]
Comment (mayeul-zama) [reply tfhe/src/shortint/ciphertext/mod.rs:121]: Yes,
I'd rather not the order as `(deg1 + deg 2).valid(max_degree)` is uglier than `max_degree.valid(deg1 + deg 2)`
However changing the name  would be fine

[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/shortint/ciphertext/mod.rs:121]: don't have a strong opinion, I misremebered the noise thing, I potentially like max_degree.validate(degree), but if we leave it as is it's fine

I'll take a closer look to the rest of the PR

File: tfhe/src/shortint/ciphertext/mod.rs [hunk 2]
```diff
@@ -102,6 +98,56 @@ impl std::ops::Mul<usize> for NoiseLevel {
     }
 }
 
+/// Maximum value that the degree can reach.
+#[derive(Debug, PartialEq, Eq, Copy, Clone, Serialize, Deserialize)]
+pub struct MaxDegree(usize);
+
+impl MaxDegree {
+    pub fn new(value: usize) -> Self {
```
[the remark this answers is not in the document]
Comment (mayeul-zama) [reply tfhe/src/shortint/ciphertext/mod.rs:106]: I don't think that's very useful
Do you think it is ?

If we ever need some functions to be `const`, it's easy to add later

File: tfhe/src/shortint/ciphertext/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tmontaigu) [reply tfhe/src/shortint/ciphertext/mod.rs:121]: I would say `validate` would be a better name

**Review by tmontaigu — APPROVED** [review]

**Pull request merged**